### PR TITLE
Remove broken KeyIterator on 0.7+ in JuMP 0.18

### PR DESCRIPTION
**Pull request by coroa**
As an alternative to #1658, we here remove the `KeyIterator` struct altogether and provide `keys` on a `JuMPArray` directly by `Iterators.product`.

As explained in the discusson of #1658, the only drawback is that `ProductIterator` has a shape.

For instance, `size(product(1:2, 3:5)) = (2, 3)`, s.t. `collect(product(1:2, 3:5))` becomes
```
julia> collect(Iterators.product(1:2, 3:5))
2×3 Array{Tuple{Int64,Int64},2}:
 (1, 3)  (1, 4)  (1, 5)
 (2, 3)  (2, 4)  (2, 5)
```
contrarily to what the current and previous `KeyIterator` returned
```
julia> collect(keys(JuMPArray(rand(2,3), (1:2, 3:5))))
6-element Array{Any,1}:
 (1, 3)
 (2, 3)
 (1, 4)
 (2, 4)
 (1, 5)
 (2, 5)
```

For its main purpose, i.e. the iteration over all index pairs, that would not have any impact, of course.

### CODE REVIEW

File: src/JuMPContainer.jl [hunk 1]
```diff
@@ -220,140 +220,3 @@ else
     Base.iterate(it::ValueIterator, k) = iterate(it.x, k)
 end
 Base.length(it::ValueIterator) = length(it.x)
-
-mutable struct KeyIterator{JA<:JuMPArray}
```
Comment (mlubin) [inline src/JuMPContainer.jl]: I wonder why we had such a complicated implementation before.

File: test/variable.jl [hunk 2]
```diff
@@ -219,6 +219,19 @@ end
...
     end
 
+    @testset "Key iterators of JuMPContainers" begin
+        m = Model()
+        @variable(m, x[2:3, 1:2])
+        @variable(m, y[2:3])
+        @variable(m, z[-1:1,[:red,"blue"]])
+        @variable(m, w[i=1:2,j=1:2 ; (j+i) % 2 == 1])
+
+        @test all(reshape(collect(keys(x)), :) .== [(2,1), (3,1), (2,2), (3,2)])
```
Comment (mlubin) [inline test/variable.jl]: Is `all` needed, or could we just check for equality of the vectors and sets?

Comment (mlubin) [inline test/variable.jl]: Could we replace `reshape(foo, :)` with `vec(foo)`?

Comment (coroa) [reply test/variable.jl]: didn't remember about vec, sure

Comment (coroa) [reply test/variable.jl]: sure

File: test/variable.jl [hunk 3]
```diff
@@ -219,6 +219,19 @@ end
...
     end
 
+    @testset "Key iterators of JuMPContainers" begin
+        m = Model()
+        @variable(m, x[2:3, 1:2])
+        @variable(m, y[2:3])
+        @variable(m, z[-1:1,[:red,"blue"]])
+        @variable(m, w[i=1:2,j=1:2 ; (j+i) % 2 == 1])
+
+        @test vec(collect(keys(x))) == [(2,1), (3,1), (2,2), (3,2)]
```
Comment (mlubin) [inline test/variable.jl]: Nit: The whitespace doesn't follow the JuMP style guide (there should be spaces after commas). However, not a big deal for the release-0.18 branch.

**Comment by mlubin** [comment]
Thanks!

**Pull request merged**